src/ql/pass/opt/dead_code_elim/dead_code_elim.cc

```diff
@@ -0,0 +1,174 @@
+                            if_else->branches.remove(i);
+                        }
+                        if_else->otherwise.reset();     // removes Maybe item
+
+                        // if we're the sole if_else-branch (remaining), turn body (sub_block) into statements
+                        if(branch_idx == 0) {
+                            QL_IOUT("turn body of sole 'if(true)' branch into statements");
+                            for (auto &st : branch->body->statements) {
+                                block->statements.add(st, stmt_idx++);  // NB: increment affects both this add and statement loop
+                            }
+
+                            // mark complete if_else statement for removal
+                            remove_if_else = true;
+                        }
+
+                        // NB: done looping, since we erased everything that followed
+                        break;
+
+                    } else {    // condition 'false'
+                        // NB: no need to descend body, since we'll discard it
+                        QL_IOUT("removing dead if-branch " << branch_idx);
+                        if_else->branches.remove(branch_idx);
+                        continue;   // loop, using same value for branch_idx
+
+                        // NB: may remove all if_else branches (which we repair later)
+                    }
+
+                // condition is not a bit_literal
+                } else {
+                    // descend body
+                    run_on_block(branch->body);
+                }
+                branch_idx++;
+            }
+
+            // descend otherwise
+            if (!if_else->otherwise.empty()) {
+                run_on_block(if_else->otherwise);
+            }
+
+            // if we no longer have branches, but do have otherwise, promote its body into statements within this block
+            if (if_else->branches.empty()) {
+                if (!if_else->otherwise.empty()) {
+                    QL_IOUT("turn body of final 'if_else->otherwise' into statements");
+
+                    // move if_else->otherwise statements
+                    for (auto &st : if_else->otherwise->statements) {
+                        block->statements.add(st, stmt_idx++);  // NB: increment affects both this add and statement loop
+                    }
+                }
+
+                remove_if_else = true;
+            }
+
+            // if needed, remove the complete if_else statement altogether
+            if (remove_if_else) {
+                block->statements.remove(if_else_idx-1);    // FIXME: why is -1 necessary?
+            }
+
+        // handle loop
+        } else if (auto loop = statement->as_loop()) {
+            // descend loop body
+            run_on_block(loop->body);
+
+            // NB: we currently have no real use for optimizing static loops, note that we cannot fully
+            // remove loop anyway if break or continue exists
+        }
```

Comment:
It would be nice to convert a C-style for loop to a foreach loop, when possible. Of a foreach loop it is easier to compute the iteration count, and that is something that passes may like to do.
But this only becomes relevant when OpenQL would unroll loops and other loop optimizations, and that is far away.

Reply:
Agree